### ISSUE TITLE
[Loopring 3.6] Split password reset from deposit

### ISSUE THREAD

**Issue by dong77** [Loopring/protocols#1242]
The idea is to make registration and password reset decoupled from onchain deposits, so the relayer can sort requests from the same user with flexilibity.

@hosschao @lydy please provide an example here.

**Comment by Brechtpd** [comment]
I want this as well. :)

I also want to make it more flexible to handle onchain requests because the fixed order they need to to be processed is too limiting. I want to make them more like how conditional transfers are processed where the operator can just pick the requests it wants to process. No more virtual chains/queues. Would that be okay?

Would it be okay to have Registration/Password Reset in a separate circuit? The problem is getting the EdDSA public key inside the circuit, it's too much data to do it similarly to other requests (so it increases the cost of all other requests quite a bit in number of constraints).

So there will probably be 2 different circuits:
- Registration/Password Reset
- Everything else (trades, transfers, onchain/offchain withdrawal, deposits)

I'll have to see how much the extra cost actually is, maybe I can still find a way to do Registration/Password Reset efficiently in the same circuit as well.

**Comment by dong77** [comment]
> I want this as well. :)
> 
> I also want to make it more flexible to handle onchain requests because the fixed order they need to to be processed is too limiting. I want to make them more like how conditional transfers are processed where the operator can just pick the requests it wants to process. No more virtual chains/queues. Would that be okay?
> 

Will it be more expensive if the operator can cherry-pick requests instead of processing them one after another in a fixed order? One benefit of the current design (virtual chains) is that the contract just need to check the oldest request's age to determine if the operator must process the request ASAP which is more efficient.  My only concern is the new approach will cost more gas.

> Would it be okay to have Registration/Password Reset in a separate circuit? The problem is getting the EdDSA public key inside the circuit, it's too much data to do it similarly to other requests (so it increases the cost of all other requests quite a bit in number of constraints).

Yes, I think the handling of EdDSA keys are very different from other circuits and it makes sense to have a circuit for registration/password reset.

BTW, are we using one key per request type? Meaning the account node will hold multiple account keys per account?
> So there will probably be 2 different circuits:
> 
> * Registration/Password Reset
> * Everything else (trades, transfers, onchain/offchain withdrawal, deposits)
> 
> I'll have to see how much the extra cost actually is, maybe I can still find a way to do Registration/Password Reset efficiently in the same circuit as well.

**Comment by Brechtpd** [comment]
> Will it be more expensive if the operator can cherry-pick requests instead of processing them one after another in a fixed order? 

I don't think so, especially because we now have to support deposits scattered throughout the block. In any case, processing onchain requests will actually have **negative gas cost**. We can now just reset storage used by these onchain requests which will provide the operator with a refund of 15,000 gas. The operator can actually use these to try and minimize the total gas cost for the block submissions.

> One benefit of the current design (virtual chains) is that the contract just need to check the oldest request's age to determine if the operator must process the request ASAP which is more efficient. My only concern is the new approach will cost more gas.

I would just use a notification system here like in my margin trading design. The contract doesn't keep track of the all the different ages of all requests, it just keeps a timestamp for the request and we allow anyone to call a function on the exchange contract that points to a request that is too old and the contract can then takes the necessary steps (and the notifier gets nice reward using the exchange stake). I think that's a nice and simple solution.

> BTW, are we using one key per request type? Meaning the account node will hold multiple account keys per account?

Just 1 keypair per account for all requests. I still think different keys for different requests doesn't add anything really. When we only had trading available on loopring.io and a user's private key is exposed, this is still a critical security risk and all funds are at risk. So it would be strange to say that it's safe to share your trading key with anyone...

Nolan (from the staking UI) is building a bot that allows trading strategies to be shared via the bot, which I think is a much more secure way to do it because no keys need to shared at all, but this comes at the cost of convenience because of course because the bot needs to be running for the user somewhere.

If people really want to give over complete control I think it would be better to just go full centralized exchange and build these restrictions into the relayer. This way the user either needs to trust the entity that has his keys or the exchange itself, so still two barriers away from somebody exploiting his keys. Arguably this is even better than having different sets of keys because the relayer can provide more fine grained control over what's allowed and what not, and the relayer won't have access to the keys so no direct reason to not actually impose those restrictions.

**Comment by dong77** [comment]
@Brechtpd given that transfers and trades are going into the same circuits, I guess it's hard to implement bitmap access control for transfers and orders, right?

**Comment by Brechtpd** [comment]
I wouldn't say it's necessarily harder, just messier. But I'm also biased because I think the benefits, if any, are really minimal for the amount of work it would take to implement this. I think the alternatives I talked about above are nice, but certainly also not perfect, solutions.

**Comment by dong77** [comment]
Maybe it's easier NOT to implement ACL but simply allow each user to enable/disable transfers on layer1 (enabled by default)? But I believe the gas price for checking that flag for all accounts in a block will be too expensive.
